ex5/y_ogawa/main.py

```diff
@@ -0,0 +1,348 @@
+        means[k, :] = np.sum(responsibilities[:, k, np.newaxis] * data, axis=0) / N_k[k]
+
+    # 共分散行列の再計算
+    for k in range(n_components):
+        # データポイントと平均の差分を計算
+        diff = data - means[k, :]
+
+        # データポイントと平均の差分を重み付けし、行列の積を計算
+        weighted_diff = np.dot((responsibilities[:, k, np.newaxis] * diff).T, diff)
+
+        # 共分散行列を計算し、N_k で割る
+        covariances[k] = weighted_diff / N_k[k]
+
+    # 混合係数の再計算
+    weights = N_k / np.sum(N_k, axis=0)
+
+    return weights, means, covariances
+
+
+def EMstep(
+    data: np.ndarray, weights: np.ndarray, means: np.ndarray, covariances: np.ndarray
+):
+    """EMアルゴリズムの実行.
+
+    Args:
+        data (np.ndarray): データ
+        weights (np.ndarray): 混合係数
+        means (np.ndarray): 平均
+        covariances (np.ndarray): 共分散行列
+
+    Returns:
+        log_likelihood (list): 対数尤度のリスト
+        weights (np.ndarray): 再計算後の混合係数
+        means (np.ndarray): 再計算後の平均
+        covariances (np.ndarray): 再計算後の共分散行列
+    """
+    n_samples = data.shape[0]
+
+    # EMアルゴリズムの試行回数
+    max_try = 100
+
+    # 対数尤度を格納する
+    log_likelihood_list = []
+
+    for j in range(max_try):
+        # Eステップを実行する
+        responsibilities, responsibilities_sum = estep(
+            data, weights, means, covariances
+        )
+
+        # Mステップを実行する
+        weights, means, covariances = mstep(data, responsibilities)
+
+        log_likelihood = 0
+        for i in range(n_samples):
+            log_likelihood += np.log(responsibilities_sum[i])
+
+        log_likelihood_list.append(log_likelihood[0])
+
+        # 収束判定
+        if j > 0 and np.abs(log_likelihood_list[-1] - log_likelihood_list[-2]) < 1e-4:
+            break
+
+    return log_likelihood_list, weights, means, covariances
+
+
+def make_likelihoodgraph(log_likelihood_list: list):
+    """再計算回数と対数尤度の関係をプロットする.
+
+    Args:
+        log_likelihood_list (list): 対数尤度
+    """
+    # 対数尤度のプロット
+    plt.plot(np.arange(len(log_likelihood_list)), log_likelihood_list)
+    plt.xlabel("Iterations")
+    plt.ylabel("Log Likelihood")
+    plt.title("Log Likelihood by Iterations")
+
+    # 横軸を整数にする
+    plt.gca().get_xaxis().set_major_locator(ticker.MaxNLocator(integer=True))
+    plt.savefig("likelihood.png")
+    plt.show()
+
+
+def plot_gmm(
+    data: np.ndarray, weights: np.ndarray, means: np.ndarray, covariances: np.ndarray
+):
+    """混合ガウス分布をプロットする.
+
+    Args:
+        data (np.ndarray): データ
+        weights (np.ndarray): 混合係数
+        means (np.ndarray): 平均
+        covariances (np.ndarray): 共分散行列
+    """
+    n_components = weights.shape[0]
+
+    # 1次元データの場合
+    if data.shape[1] == 1:
+        x = np.linspace(data.min(), data.max(), 1000)
+        y = np.zeros_like(x)
+
+        # 混合ガウス分布の確率密度関数を計算
+        for k in range(n_components):
+            y += weights[k] * multivariate_normal.pdf(
+                x, mean=means[k], cov=covariances[k]
+            )
+        zeros = np.zeros(len(data))
+
+        # 確率密度関数をプロット
+        plt.plot(x, y)
```

Comment:
```suggestion
        plt.plot(x, y, color="b", label="GMM")
```
混合ガウス分布にもラベルをつけておきましょう